### ISSUE TITLE
Prohibit leading and trailing whitespace in field names before releasing 5.2.0

### ISSUE THREAD

**Issue by sobjornstad** [Jermolene/TiddlyWiki5#5964]
Field names with leading and/or trailing whitespace (e.g., ` myfield`, `myfield    `) should be trimmed before creating the field through any mechanism. **We have a brief window of opportunity to fix this now** before we release 5.2.0, because spaces of any kind are prohibited in field names currently. If we make the change now, we won't break backwards compatibility even in pathological cases.

Reasons why leading and trailing whitespace should be prohibited:

* They are almost impossible to distinguish within the GUI, since there is already whitespace before and after the name when it's displayed.
* It's hard to see why a user would intentionally add such space, except maybe to integrate with another system which allowed it for some reason. Once someone has accidentally added the space, it's extremely difficult to figure out what happened when, e.g., filters against that field don't work.
* If we continue to prohibit leading and trailing whitespace, I think we would be able to allow whitespace in between filter steps in the future, lack of which is a common annoyance for new users (but maybe I'm missing something there). If leading whitespace becomes allowable, e.g., ` links[]` would become a valid field name we might theoretically want to filter on.
* Overall, leading and trailing whitespace in field names seems much more confusing than helpful. Unix filenames have [suffered enormously](https://dwheeler.com/essays/fixing-unix-linux-filenames.html) from being overly permissive in what characters they allow, including leading and trailing whitespace; my instinct tells me this is likely to be a continuing irritation.

In the 5.2.0 prerelease, the GUI already trims the names of such fields, but you can import tiddlers that contain such fields, or create them with an action or `$edit-text` widget. These mechanisms (and any others I'm not thinking of) should also perform the trimming.

The one trick I foresee here is, what do we do if importing a file that contains fields that differ only in whitespace, e.g.,:

```json
[
    {
        "title": "Test Tiddler",
        "description": "Value 1",
        " description": "Value 2"
    }
]
```

I don't know what we do currently if a field is duplicated in an imported tiddler – probably we could just do the same thing here.

See https://groups.google.com/g/tiddlywiki/c/l9eq2MzlRWw for previous discussion.

**Comment by saqimtiaz** [comment]
I am not against the change in principle but note #4410 which suggests that backwards compatibility would already be a concern.

**Comment by sobjornstad** [comment]
@saqimtiaz: Good catch! But as @kookma said on that thread:

> Also, when you save and reload the file, the upper case letters are automatically re-mapped to lowercase, but the spaces result in TWO fields (e.g., "this field" becomes two fields, "hi" and "test").

I tested on 5.1.23 and confirmed that leading/trailing space is stripped during this same process on save/reload. So while it's possible to *temporarily* get a field with leading/trailing space in the current version due to a bug, I don't think anyone could plausibly be relying on that behavior since they don't survive a save.

**Comment by Jermolene** [comment]
> Field names with leading and/or trailing whitespace (e.g., ` myfield`, `myfield `) should be trimmed before creating the field through any mechanism. **We have a brief window of opportunity to fix this now** before we release 5.2.0, because spaces of any kind are prohibited in field names currently. If we make the change now, we won't break backwards compatibility even in pathological cases.

At the moment, field names are trimmed by the `<$fieldmangler>` widget when handling the `tm-add-field` message. That's a high level API; it's possible to use the lower level `<$action-setfield>` widget to create tiddlers with fields starting or ending with whitespace.

The lowest level point at which we could trim field names is in the underyling `store.addTiddler()` method (the only way to create or update tiddlers is with the `store.addTiddler()` method).

However, it has two important aspects to its semantics which rule out trimming fieldnames at this level:

* The function cannot fail with an error
* The function stores the tiddler exactly as specified: storing and retrieving the same tiddler is guaranteed to return the same values

So, I think the best we can do is to trim fieldnames in the UI.

> I tested on 5.1.23 and confirmed that leading/trailing space is stripped during this same process on save/reload. So while it's possible to _temporarily_ get fields with leading/trailing space in the current version due to a bug, I don't think anyone could plausibly be relying on that behavior since they don't survive a save.

Fieldnames with leading or trailing whitespace work OK under Node.js, it's only the round-tripping through the HTML saver that fails. There are a lot of applications of TW that don't use a saver.

**Comment by sobjornstad** [comment]
> So, I think the best we can do is to trim fieldnames in the UI.

I was going to say couldn't we just prohibit them altogether, but after a test, it appears that currently `$action-setfield` lets you create fields with completely invalid names. So I guess this isn't a new problem.

> Fieldnames with leading or trailing whitespace work OK under Node.js, it's only the round-tripping through the HTML saver that fails. There are a lot of applications of TW that don't use a saver.

Are you sure about this? My test was on Node.js on 5.1.23. I just created the tiddlers, accepted the changes, and reloaded the page, no HTML saver involved. (Though perhaps this is irrelevant if your conclusion is there's nothing we can do.)

**Comment by Jermolene** [comment]
> Are you sure about this? My test was on Node.js on 5.1.23. I just created the tiddlers, accepted the changes, and reloaded the page, no HTML saver involved. (Though perhaps this is irrelevant if your conclusion is there's nothing we can do.)

I tested things by adding to a wiki folder a JSON tiddler with containing a fieldname starting with a space, and then building the wiki, running it in the browser and checking via the browser console that the whitespace was intact.

**Comment by saqimtiaz** [comment]
> I tested on 5.1.23 and confirmed that leading/trailing space is stripped during this same process on save/reload. So while it's possible to _temporarily_ get fields with leading/trailing space in the current version due to a bug, I don't think anyone could plausibly be relying on that behavior since they don't survive a save.

Not quite. It is entirely possible to have a workflow that generates and relies on a $:/temp (or $:/state) tiddler that never gets saved but contains such fields.

**Comment by sobjornstad** [comment]
> I tested things by adding to a wiki folder a JSON tiddler with containing a fieldname starting with a space, and then building the wiki, running it in the browser and checking via the browser console that the whitespace was intact.

Ah, I made my change within the wiki. So it must happen on the sync back to the server, that makes sense.

> Not quite. It is entirely possible to have a workflow that generates and relies on a $:/temp (or $:/state) tiddler that never gets saved but contains such fields.

Fair enough.
